File: packages/engine-runner/src/lobium-capabilities.ts
import { execFile } from 'node:child_process';
import { stat } from 'node:fs/promises';
import { promisify } from 'node:util';
import type { FingerprintLaunchPolicy } from '@lobster/shared-types';

const execFileAsync = promisify(execFile);

export const LOBIUM_CAPABILITY_SWITCH = '--lobium-fingerprint-capabilities';
export const LOBIUM_CAPABILITY_CONTRACT_VERSION = 1;

/**
 * Mirror of the list in `lobium/src/lobium_capabilities.cc`, which is the single source of truth —
 * it sits beside the hooks it describes, so it cannot claim a hook that was never compiled.
 *
 * `ci/validation/patch-series.test.mjs` fails the build if the two lists diverge. This copy exists
 * only so TypeScript can type-check capability names; it is never the authority on what a binary
 * actually contains — that comes from probing the binary itself.
 */
export const LOBIUM_NATIVE_FINGERPRINT_CAPABILITIES = [
  'config-channel-v1',
  /** UA, UA-CH brands, platform, hardwareConcurrency, deviceMemory and maxTouchPoints. */
  'navigator-ua-ch',
  'navigator-languages',
  'network-accept-language',
  'process-locale-timezone',
  'native-geolocation',
  'webrtc-policy',
  'webgl-deep',
  'webgl2-deep',
  'screen-metrics',
  /** Compiled everywhere, but only meaningful for an Android persona, so never required on desktop. */
  'mobile-persona',
  'canvas-farbling',
  'webgl-farbling',
  'audio-farbling',
  'client-rects',
  'media-devices',
  'webgpu-adapter',
  'native-timezone',
  /** Windows-only: the engine reports it just on win64 builds, so never require it elsewhere. */
  'font-isolation',
] as const;

export type LobiumNativeFingerprintCapability =
  (typeof LOBIUM_NATIVE_FINGERPRINT_CAPABILITIES)[number];

export interface LobiumBuildCapabilities {
  contractVersion: number;
  product: 'Lobium';
  capabilities: LobiumNativeFingerprintCapability[];
}

const KNOWN_CAPABILITIES = new Set<string>(LOBIUM_NATIVE_FINGERPRINT_CAPABILITIES);
const probeCache = new Map<string, Promise<LobiumBuildCapabilities>>();

function parseCapabilities(stdout: string, executablePath: string): LobiumBuildCapabilities {
  let value: unknown;
  try {
    value = JSON.parse(stdout.trim());
  } catch {
    throw new Error(
      `Lobium binary ${executablePath} did not return a valid native capability manifest`,
    );
  }
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error(`Lobium binary ${executablePath} returned an invalid capability manifest`);
  }
  const manifest = value as Record<string, unknown>;
  if (
    manifest.contractVersion !== LOBIUM_CAPABILITY_CONTRACT_VERSION ||
    manifest.product !== 'Lobium' ||
    !Array.isArray(manifest.capabilities) ||
    manifest.capabilities.some(
      (capability) => typeof capability !== 'string' || !KNOWN_CAPABILITIES.has(capability),
    )
  ) {
    throw new Error(
      `Lobium binary ${executablePath} has an incompatible native capability contract`,
    );
  }
  return {
    contractVersion: LOBIUM_CAPABILITY_CONTRACT_VERSION,
    product: 'Lobium',
    capabilities: [...new Set(manifest.capabilities)] as LobiumNativeFingerprintCapability[],
  };
}

/**
 * Ask the exact executable that will be spawned which native hooks it contains. An old/unpatched
 * Chromium binary either times out or returns no manifest and is rejected; filename/version claims are
 * deliberately not trusted.
 */
export async function probeLobiumBuildCapabilities(
  executablePath: string,
): Promise<LobiumBuildCapabilities> {
  const info = await stat(executablePath);
  const cacheKey = `${executablePath}:${info.size}:${info.mtimeMs}`;
  let pending = probeCache.get(cacheKey);
  if (!pending) {
    pending = execFileAsync(executablePath, [LOBIUM_CAPABILITY_SWITCH], {
      encoding: 'utf8',
      timeout: 5_000,
      maxBuffer: 64 * 1024,
      windowsHide: true,
    })
      .then(({ stdout }) => parseCapabilities(stdout, executablePath))
      .catch((error: unknown) => {
        probeCache.delete(cacheKey);
        const detail = error instanceof Error ? error.message : String(error);
        throw new Error(
          `cannot prove native fingerprint capabilities for ${executablePath}: ${detail}`,
        );
      });
    probeCache.set(cacheKey, pending);
  }
  return pending;
}

export function requiredLobiumCapabilities(
  policy: FingerprintLaunchPolicy,
  hasConfiguredGeolocation: boolean,
  platform: NodeJS.Platform = process.platform,
): LobiumNativeFingerprintCapability[] {
  const required: LobiumNativeFingerprintCapability[] = [
    'config-channel-v1',
    // navigator.userAgent, the Sec-CH-UA family, navigator.platform, hardwareConcurrency and
    // deviceMemory. Unconditional, and the most important entry in this list: a build that has the
    // config channel but not this hook accepts the persona, launches successfully, and then reports
    // the HOST's identity on the surfaces every detector reads first — a failure that looks exactly
    // like a working profile until the account is banned.
    'navigator-ua-ch',
    'navigator-languages',
    'network-accept-language',
    'process-locale-timezone',
    'webrtc-policy',
    'webgl-deep',
    // Required alongside webgl-deep, not instead of it: a build with only the WebGL1 hooks lets a
    // WebGL2 context report the host's extension list and component limits while WebGL1 reports the
    // persona's, so the two contexts disagree on one page. That is worse than neither being spoofed.
    'webgl2-deep',
    // screen.*, devicePixelRatio and the CSS device-size media values. Unconditional: every persona
    // claims a display, and an unspoofed screen block contradicts the rest of the persona outright.
    'screen-metrics',
    'media-devices',
    // navigator.gpu names the same GPU as WEBGL_debug_renderer_info. Unconditional for the same
    // reason: an unhooked WebGPU adapter reports the real card next to a spoofed WebGL renderer.
    'webgpu-adapter',
    // Applied inside the engine because the TZ environment variable is POSIX-only. On Windows the
    // process-locale route is a no-op, so without this hook the persona timezone silently does not
    // apply at all — the failure mode that made this a required capability rather than an optional
    // one.
    'native-timezone',
  ];
  if (hasConfiguredGeolocation) required.push('native-geolocation');
  // Windows resolves fonts through DirectWrite in the browser process; Linux and macOS reach the
  // same isolation through the launcher's per-profile FONTCONFIG_FILE, which is not a property of
  // the binary. Requiring it everywhere would fail launches on platforms that never compile it.
  if (platform === 'win32') required.push('font-isolation');
  if (policy.hardwareNoise.canvas) required.push('canvas-farbling');
  if (policy.hardwareNoise.webgl) required.push('webgl-farbling');
  if (policy.hardwareNoise.audio) required.push('audio-farbling');
  if (policy.hardwareNoise.clientRects) required.push('client-rects');
  return required;
}

export function assertLobiumBuildCapabilities(
  manifest: LobiumBuildCapabilities,
  required: readonly LobiumNativeFingerprintCapability[],
): void {
  const available = new Set(manifest.capabilities);
  const missing = required.filter((capability) => !available.has(capability));
  if (missing.length > 0) {
    throw new Error(`Lobium build lacks required native fingerprint hooks: ${missing.join(', ')}`);
  }
}

File: lobium/src/lobium_capabilities.cc
// Copyright 2026 The Lobster Browser Authors.
//
// See lobium_capabilities.h.

#include "components/lobium_fp/lobium_capabilities.h"

#include <string_view>
#include <vector>

#include "base/strings/strcat.h"
#include "base/strings/string_number_conversions.h"
#include "build/build_config.h"

namespace lobium {

namespace {

// Hooks compiled on every platform.
//
// Each name corresponds to a patch in lobium/patches/series; the mapping is documented per hook in
// lobium/hooks.md and enforced by ci/validation/patch-series.test.mjs.
constexpr std::string_view kPortableCapabilities[] = {
    // The config channel itself. Everything below is inert without it, so a build that reports any
    // capability must report this one.
    "config-channel-v1",
    // navigator.userAgent/platform/hardwareConcurrency/deviceMemory/maxTouchPoints, the outgoing
    // User-Agent header and the whole Sec-CH-UA family. Reported separately from the config channel
    // because the two are genuinely separable: core/navigator-ua-ch.patch hooks five upstream files
    // that core/config-channel.patch does not touch, so a tree with the transport applied and this
    // patch rejected still parses the persona and still reports the HOST on every one of those
    // surfaces - a failure that looks exactly like a working launch from the outside.
    "navigator-ua-ch",
    // navigator.webdriver is a separate Blink hook. Keeping it separate prevents a build that has
    // the UA/UA-CH patch but missed navigator-webdriver.patch from claiming automation protection.
    "navigator-webdriver",
    "navigator-languages",
    "network-accept-language",
    "process-locale-timezone",
    "native-geolocation",
    "webrtc-policy",
    // WebGL1: vendor/renderer, VERSION, SHADING_LANGUAGE_VERSION, MAX_* caps, shader precision, and
    // the registration-ordered extension list.
    "webgl-deep",
    // WebGL2: the per-context extension list, plus the uniform/varying COMPONENT limits derived from
    // the persona's vector limits so the two contexts cannot contradict each other.
    "webgl2-deep",
    // screen.*, window.devicePixelRatio, and the CSS device-size media values, which must agree with
    // each other and with the persona's claimed display.
    "screen-metrics",
    // Android persona parity in Blink: suppression of the desktop PDF plugin surface when the config
    // declares uaMobile. Only required for an Android persona, but always compiled.
    //
    // NARROWED 2026-08-26. This used to read "Touch points, pointer/hover media features and the
    // rest of the mobile-shaped surfaces", which over-claimed by a wide margin:
    // fingerprint/mobile-persona.patch hooks exactly one upstream file,
    // third_party/blink/renderer/modules/plugins/dom_plugin_array.cc, and nothing else.
    //
    // Where those surfaces actually come from:
    //   * navigator.maxTouchPoints -> navigator-ua-ch, which reports it with the rest of navigator.
    //   * (pointer: coarse) / (hover: none) -> NOT from the binary at all. They come from the CDP
    //     Emulation.setDeviceMetricsOverride{mobile:true} that the Android path installs
    //     (packages/engine-runner/src/mobile-emulation.ts).
    //
    // Measured on the Windows host: an Android persona launched with only --lobium-fp-config - the
    // desktop path, no CDP emulation - reports uaMobile true, maxTouchPoints 5 and a 393x873 screen
    // while (pointer: coarse) and (hover: none) both answer FALSE. Over-reporting is the dangerous
    // direction for this contract: the sidecar requires mobile-persona for a mobile launch, and if
    // anyone trusted the old wording and dropped the CDP layer, pointer and hover would silently
    // revert to desktop next to an Android UA.
    "mobile-persona",
    "canvas-farbling",
    "webgl-farbling",
    "audio-farbling",
    "client-rects",
    "media-devices",
    // navigator.gpu adapter identity, derived from the same GPU the WebGL renderer names.
    "webgpu-adapter",
    // The persona timezone applied inside the engine. Distinct from process-locale-timezone: the TZ
    // environment variable is POSIX-only, and on Windows ICU reads the registry instead, so the
    // env-var route silently does nothing there.
    "native-timezone",
};

}  // namespace

std::string CapabilityManifestJson() {
  std::vector<std::string_view> names(std::begin(kPortableCapabilities),
                                      std::end(kPortableCapabilities));

#if BUILDFLAG(IS_WIN)
  // Native Windows font-isolation machinery: one restricted DirectWrite collection shared by
  // FontDataService, character fallback, local(), Local Font Access, chrome.fontSettings, and the
  // legacy proxy, plus verified-pack transport and CSS claimed-family alias plumbing.
  //
  // This capability says the binary has those hooks. It does NOT claim that a pack was supplied at
  // this launch, that every claimed family has a metric-exact substitute, or that physical pack
  // faces fabricate proprietary PostScript/full names in Local Font Access. Runtime gates prove
  // pack registration and report alias residuals separately.
  //
  // Windows-only on purpose, and deliberately conservative. The FontDataService half also compiles
  // on Linux (kFontDataServiceLinux is enabled by default there too), so a Linux build does carry
  // part of this. But Linux isolates fonts through the launcher's per-profile FONTCONFIG_FILE, which
  // is not a property of the binary at all, and claiming the capability there would assert coverage
  // the binary alone cannot provide. Under-reporting costs nothing — requiredLobiumCapabilities only
  // asks for it on win32 — while over-reporting would let the sidecar launch believing in isolation
  // that is not there.
  names.push_back("font-isolation");
#endif

#if BUILDFLAG(IS_LINUX) || BUILDFLAG(IS_WIN)
  // The Android phone/tablet stage: LobiumDeviceFrameView is constructed by BrowserView and kept
  // locked to the renderer emulation scale.
  //
  // This is a UI feature in a fingerprint contract on purpose. An Android persona whose window is
  // a desktop-shaped rectangle at desktop dimensions contradicts the persona it just claimed, so
  // the frame is a fingerprint surface in effect even though it is drawn rather than reported.
  //
  // It is here because of how its absence failed. The launcher emits --lobium-device-frame
  // unconditionally for a mobile profile, Chromium ignores switches it does not know, and the
  // capability contract did not cover the feature - so a binary built without it launched, reported
  // success, and simply had no frame. That shipped and went unnoticed for days. Declaring it here
  // means such a binary now refuses a mobile launch instead of quietly degrading it.
  //
  // Guarded to the platforms whose BrowserView call sites are compiled: macOS has none, so a macOS
  // build must not claim it. requiredLobiumCapabilities only asks for it on an emulated Android
  // launch, so under-reporting elsewhere costs nothing.
  names.push_back("device-frame");
#endif

  std::string out = base::StrCat(
      {"{\"contractVersion\":",
       base::NumberToString(kCapabilityContractVersion),
       ",\"product\":\"Lobium\",\"capabilities\":["});
  for (size_t i = 0; i < names.size(); ++i) {
    if (i > 0) {
      out += ',';
    }
    // The names are compile-time literals from this file, so no JSON escaping is needed - and any
    // name requiring escaping would fail the CI cross-check against the TypeScript mirror first.
    base::StrAppend(&out, {"\"", names[i], "\""});
  }
  out += "]}";
  return out;
}

}  // namespace lobium
